src/main/java/org/kuali/kfs/coa/document/validation/impl/AccountGlobalRule.java

```diff
@@ -863,5 +962,1065 @@ private boolean isIncomeStreamAccountRequired(String fundGroupCode, String subFu
+		boolean success = true;
+		if (expectFilled) {
+			success = !newAccount.getActiveIndirectCostRecoveryAccounts().isEmpty();
+			if (success) {
+				for (IndirectCostRecoveryAccountChange account : newAccount.getActiveIndirectCostRecoveryAccounts()) {
+					success &= StringUtils.isNotBlank(account.getIndirectCostRecoveryAccountNumber()) && StringUtils.isNotBlank(account.getIndirectCostRecoveryFinCoaCode());
+				}
+			}
+
+			if (!success) {
+				acctGlobalDetail.refreshReferenceObject(KFSPropertyConstants.ACCOUNT);
+
+				success &= !acctGlobalDetail.getAccount().getActiveIndirectCostRecoveryAccounts().isEmpty();
+			}
+		}
+
+		if (!expectFilled) {
+			success = newAccount.getActiveIndirectCostRecoveryAccounts().isEmpty();
+			if (success) {
+				acctGlobalDetail.refreshReferenceObject(KFSPropertyConstants.ACCOUNT);
+				success &= acctGlobalDetail.getAccount().getActiveIndirectCostRecoveryAccounts().isEmpty();
+			}
+		}
+
+		if (!success) {
+			putFieldError(KFSPropertyConstants.INDIRECT_COST_RECOVERY_ACCOUNTS,errorMessage, args);
+		}
+		
+		return success;
+    }
+    
+    /**
+     * This method is a helper method that replaces error tokens with values for contracts and grants labels
+     *
+     * @param errorConstant
+     * @return error string that has had tokens "{0}" and "{1}" replaced
+     */
+    protected String replaceTokens(String errorConstant) {
+        String cngLabel = getSubFundGroupService().getContractsAndGrantsDenotingAttributeLabel();
+        String cngValue = getSubFundGroupService().getContractsAndGrantsDenotingValueForMessage();
+        String result = getConfigService().getPropertyValueAsString(errorConstant);
+        result = StringUtils.replace(result, "{0}", cngLabel);
+        result = StringUtils.replace(result, "{1}", cngValue);
+        return result;
+    }
+    
+    /**
+     * This method checks to make sure that if the contract control account exists it is the same as the Account that we are working
+     * on
+     *
+     * @param newAccount
+     * @return false if the contract control account is entered and is not the same as the account we are maintaining
+     */
+    protected boolean checkContractControlAccountNumberRequired(CuAccountGlobal newAccount) {
+
+        boolean result = true;
+
+        // Contract Control account must either exist or be the same as account being maintained
+
+        if (ObjectUtils.isNull(newAccount.getContractControlFinCoaCode())) {
+            return result;
+        }
+        if (ObjectUtils.isNull(newAccount.getContractControlAccountNumber())) {
+            return result;
+        }
+        
+        //if no account global details exist then don't validate
+        if(ObjectUtils.isNull(newAccount.getAccountGlobalDetails()) || (ObjectUtils.isNotNull(newAccount.getAccountGlobalDetails()) && newAccount.getAccountGlobalDetails().size() == 0)){
```

Comment:
This logic is effectively "x || (!x && y)"; I would recommend rewriting as "x || y" instead.